src/framework/utils/debug-draw.js

```diff
@@ -0,0 +1,261 @@
+Object.assign(pc, function () {
+
+    // DebugLine object represents an array of point segments together with their colors, and some additional properties
+    function DebugLine(points, colors, endTime, id, options) {
+        this.points = points;
+        this.colors = colors;
+        this.endTime = endTime;
+        this.id = id;
+        this.isAlive = true;
+
+        // default options if not supplied
+        this.options = options || {};
+        if (typeof this.options.layer === 'undefined')
+            this.options.layer = pc.app.scene.layers.getLayerById(pc.LAYERID_IMMEDIATE);
+        if (typeof this.options.depthTest === 'undefined')
+            this.options.depthTest = true;
+    }
+
+    var _currentTime = 0;
+    var _deltaTime = 0.1;   // the default for editor as it's not updated there
+    var lines = [];
+
+    var render = function () {
+        _currentTime += _deltaTime;
+        var needDeleteLines = false;
+
+        // render lines
+        var i;
+        for (i = 0; i < lines.length; i++) {
+            if (lines[i].isAlive) {
+                pc.app._addLines(lines[i].points, lines[i].colors, lines[i].options);
```

Comment:
App doesn't have an _allLines function... what am I missing?!

Reply:
it's in immediate.js

Reply:
Eeeeuw immediate assigns to Application.prototype! (mindblownemoji)

Reply:
Yeah, this could definitely do with being tidied up. I don't think it's necessary for this PR to address this problem though. But @mvaligursky, maybe add an issue to refactor this part of the engine?